sympy/stats/frv_types.py

```diff
@@ -259,18 +256,12 @@ def Coin(name, p=S.Half):
 class BinomialDistribution(SingleFiniteDistribution):
     _argnames = ('n', 'p', 'succ', 'fail')
 
-    def __new__(cls, *args):
-        n = args[BinomialDistribution._argnames.index('n')]
-        p = args[BinomialDistribution._argnames.index('p')]
-        n_sym = sympify(n)
-        p_sym = sympify(p)
-
-        if fuzzy_not(fuzzy_and((n_sym.is_integer, n_sym.is_nonnegative))):
-            raise ValueError("'n' must be positive integer. n = %s." % str(n))
-        elif fuzzy_not(fuzzy_and((p_sym.is_nonnegative, (p_sym - 1).is_nonpositive))):
-            raise ValueError("'p' must be: 0 <= p <= 1 . p = %s" % str(p))
-        else:
-            return super(BinomialDistribution, cls).__new__(cls, *args)
+    @staticmethod
+    def check(n, p, succ, fail):
+        _value_check((n.is_integer and n.is_nonnegative),
```

Comment:
`_whenFalse(fuzzy_and((n.is_integer, n.is_nonnegative)), "should be nonnegative integer: %s" % n)`

Reply:
Actually, we are only interested in the case when this evaluates to False in an obvious way so building this with operators instead of assumptions is the way to go. I think the only thing extra needed to handle the integer assumption is
`is_integer = lambda x: Piecewise((Eq(floor(x), x),x>0),(Eq(-floor(-x),x), True))`. So this query could be
```python
_whenFalse(And(is_integer(n), n>=0), "expecting nonnegative integer, but got %s" % n)
```
There is already some built-in smarts to relationals:
```
>>> var('n', odd=True, positive=True, prime=True)
n
>>> n>0
True
>>> n>1
True
>>> n>2
True
>>> n>3
n > 3
```

Reply:
i will make the changes may be by tomorrow. thanks for the reviews.

Reply:
regarding previous...I think `_whenFalse(fuzzy_and([n.is_integer, n >= 0]), ...)` is fine. The rule is:
if `a` or `b` might be None, don't use `a and b` or And(a, b). Or, as you were doing (but in too many cases, I think), use `a != False` if `a` might be None. You don't have to do that if a is a Relational, however, because if a decision can't be made it remains a Relational (which is bool True) and won't confuse the `_whenFalse` tester.

Reply:
I don't think Relational is bool true, I am getting the following error,
```
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> var('n', odd=True, positive=True, prime=True)
n
>>> n>0
True
>>> n>1
True
>>> n>2
True
>>> n>3
n > 3
>>> n>3 == True
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gagandeep/sympy/sympy/core/relational.py", line 305, in __nonzero__
    raise TypeError("cannot determine truth value of Relational")
TypeError: cannot determine truth value of Relational

```

Reply:
Can we apply the below mentioned principles:

Let's say there are two conditions, `cond1` and `cond2`, the following should be done,

1. If `cond1` or `cond2` might be `None`, then use `fuzzy_and` and `fuzzy_not`.
2. If `cond1` and `cond2` are both determined to be either `True` or `False` and nothing else then use `And` or `and`.

What do you say @smichr? Can I go with the above rules during my summer project?

Reply:
`n > 3` should be parenthesized:
```
>>> n = symbols('n')
>>> (n > 3) == True
False
```

Reply:
So is Relational, bool True, or bool False or None?

Reply:
Or none of those (in which case it is not equal to True). If you write `n > 3 == True`, Python will interpret it as `n > 3 and 3 == True` and therefore needs to evaluate `n > 3` as a boolean.

Reply:
So what to use, in such cases. `And` or `fuzzy_and`?

Reply:
Perhaps it would be possible to use two checks with different error messages: `_value_check(n.is_integer, "... expecting integer...")` and similarly for `n.is_nonnegative`.

Reply:
Please clarify the specific case you are referring to.

And note: `is_foo` attributes will be True, False or None while Relational will be True, False or Relational.

Reply:
I think that's the best way, consider the current situation. I will re-write the checks for this PR by avoiding `And`, `fuzzy_and`. Thanks for helping me out with this confusion. @smichr @Upabjojr I am going with this suggestion.

Reply:
> Please clarify the specific case you are referring to.

Like, if p > 0 and p < 1 are undetermined, I mean p can be symbols, so what to use in such cases, `And` or `fuzzy_and`?

Reply:
Or, as in the diff: 
```python
_value_check(And(n>=0, n.is_Integer), 'expecting literal nonnegative integer')
_value_check(fuzzy_and([n>=0, n.is_integer]), 'expecting nonnegative integer')
```
We could dispence with any sort of wrapper by making the `_value_check signature be 
```
def _value_check(cond, msg):
    if not iterable(cond):
        cond = [cond]
    if fuzzy_and(cond) == False:
            raise ValueError(msg)
    return True
```

Reply:
@smichr I think we can work on `_value_check` on a different PR in future because, according to me the function currently doesn't take many cases into account and few them you have suggested in your comment.